spec/index.bs

```diff
@@ -596,24 +596,36 @@ Issue(gpuweb/gpuweb#296): Consider merging all read-only usages.
 Textures may consist of separate [=mipmap levels=] and [=array layers=],
 which can be used differently at any given time.
 Each such <dfn dfn>subresource</dfn> is uniquely identified by a
-[=texture=], [=mipmap level=], and
-(for {{GPUTextureDimension/2d}} textures only) [=array layer=].
+[=texture=], [=mipmap level=],
+(for {{GPUTextureDimension/2d}} textures only) [=array layer=],
+and [=aspect=].
 
-The **main usage rule** is that any [=subresource=]
-at any given time can only be in either:
+<dfn dfn>atom resource</dfn> means a whole buffer or a single subresource of a texture.
+
+The **main usage rule** is that any [=atom resource=]
+at any given time within [=usage scope=] can only be in either:
   - a combination of [=read-only usage=]s
   - a single [=mutating usage=]
 
+The only exception is that a combination of writeonly-storage-texture usages, or a combination
```

Comment:
this isn't an exception, it falls under the main usage rule, since render pass makes a single usage scope

Reply:
I think the wording (a single [=mutating usage=]) is not clear. A single usage might mean exactly one use. "A single type of usage" can include multiple uses (but all of them are the same type), per my understanding.

Reply:
I think it is clear if you read it in the context, i.e "any subresource .... within usage scope ... can be in ... a single mutating usage". If we start talking about "single type of usage", we'd need to define what a "type of usage" is. Wouldn't it make things more complicated?

I understand what clarity issue you are trying to address. I'm just not totally sure if it needs to be reworded, and what that wording would look like.

In any case, the statement about this being an exception is incorrect.

Reply:
OK. If you think it is good, I have no problem. Forgive my English... Maybe I can add a note for this? It is like an explanation for a special case.

Reply:
Hmm, good suggestion, Kai. One question: is it OK to render into the same subresource when we render into multiple render targets (colorAttachments point to the same subresource)? It seems that race condition to storage buffer or write-only storage texture in render pass is OK, but it is not valid for render target? And race condition upon storage buffer/texture of the same atom resource in compute pass is not valid? Did you discuss this before and got consensus that "multiple instances of the same mutating usage" is valid? If not, maybe we need to discuss and clarify this validation rule.

Reply:
wow that's a great point! We haven't really discussed this, it's worth a separate issue.

Reply:
Great point. I think what we may want to do here is actually change the rule to track a list (multiset) of usages, rather than the (non-multi)set of usages, and then actually introduce the exceptions:

- multiple instances of writeonly-storage-texture in a scope
- multiple instances of storage-buffer in a scope

and also double check that we actually want to allow multiple instances of these within a single draw call or single dispatch call.

Before we figure that out, I think we should go ahead with the revisions I suggested (which should be really quick and clarify what was already written).

Reply:
Hmm, I was thinking about that differently. So supposing your suggestion is (1), here are the other 2:

(2) just specify `beginRenderPass` in a way that there has to be no collisions between subresources used for different attachments

(3) for each usage, we can have a property of whether it's "exclusive" or not. The only "exclusive" usage is OUTPUT_ATTACHMENT, I think? Exclusiveness means that multiple uses of the same resource with this usage can't be combined together.

Reply:
(3) is almost the same as (1) except that we give a name to the exceptions.

I don't have a strong preference between any of the three. The biggest problem I have with the whole rule is that it feels very arbitrary. It's not the rule we would write if trying to prevent all R/W and W/W conflicts (that would be the multiset version). I don't think there's a fundamental reason that conflicts within one usage flag are more okay than conflicts between multiple, except that we decided algorithms are more likely to need those conflicts. For example, why do we allow R/W conflicts between storage-buffer and storage-buffer but not between readonly-storage-buffer and storage-buffer? I think we need a more principled approach here.

Reply:
Discussed it on the editors meeting, decided that 1) this is out of scope of this PR, we'll address this in the next set of changes with #745, and 2) the way we'll try to describe the usage conflicts is by defining sets of "compatible" usages.

Reply:
For now we can just leave this ~as-was, with an `Issue(gpuweb/gpuweb#745):`

Reply:
> 2) the way we'll try to describe the usage conflicts is by defining sets of "compatible" usages.

So `OUTPUT_ATTACHMENT` will not be compatible with itself, while `STORAGE_BUFFER` will be?

Reply:
in simple words, yes